test/test_littlefs.c

```diff
@@ -1047,6 +1052,12 @@ TEST_CASE("Rewriting file frees space immediately (#7426)", "[littlefs]")
     test_teardown();
 }
 
+#if 0
```

Comment:
@igrr I was wondering if you or someone else at espressif could help me with this unit test? It works fine on esp-idf v4.X, but hangs on v5.0.

If you uncomment the print statement below, you get the following (on loop) after a while:

```
/IDF/tools/unit-test-app/components/littlefs/src/littlefs/lfs.c:560:error: No more free space 165
used: 520192 total: 524288
/IDF/tools/unit-test-app/components/littlefs/src/littlefs/lfs.c:560:error: No more free space 165
used: 520192 total: 524288
/IDF/tools/unit-test-app/components/littlefs/src/littlefs/lfs.c:560:error: No more free space 165
used: 520192 total: 524288
/IDF/tools/unit-test-app/components/littlefs/src/littlefs/lfs.c:560:error: No more free space 165
used: 520192 total: 524288
/IDF/tools/unit-test-app/components/littlefs/src/littlefs/lfs.c:560:error: No more free space 165
used: 520192 total: 524288
/IDF/tools/unit-test-app/components/littlefs/src/littlefs/lfs.c:560:error: No more free space 165
```

Was wondering if there's an obvious API change in v5.0 that I didn't account for properly in this PR.

Reply:
Okay, will take a look. I don't immediately see what is the issue, need to spend some time to understand what this test does.

Reply:
Basically it writes a bulk of data to get the filesystem near-full. Then, it writes one byte at a time until it cannot anymore. Apparently in some situations this could lead to upstream LFS reporting `used_bytes > total_bytes`, which is undesirable.

In v4.X, this test runs to completion (at some point `fwrite` returns either 0 or an error code, as expected). However, in v5.0 it just runs forever (I've stopped it after 10 minutes).

Reply:
@BrianPugh I've looked into the issue for a bit, here's what I found.

When running the test with IDF v5.0, the `FILE* f` used in the test gets initialized with a 4096 byte internal buffer.
When running the test with IDF v4.4, the buffer is 128 bytes.

This results in the difference in the test behavior:
With v5.0, `fwrite` function writes around 200 bytes so it doesn't fill the buffer completely. The actual write to disk (via vfs_littlefs_write) doesn't happen, since the buffer at stdio level (inside FILE structure) is not full. At this point, no error is reported, so fwrite returns with success. When fclose is called, only then does the buffer get flushed. At this point the "out of space" error is detected. Since the test doesn't check `fclose` return value, the error doesn't get detected, and the test continues running.

With v4.4, `fwrite` function writes around 200 bytes and fills the 128 buffer, so vfs_littlefs_write gets called and the error is detected immediately.

So, why does the file buffer get set to 128 bytes in v4.4 and to 4096 bytes in v5.0? It turns out to be a long story.

In ESP32, part of the newlib (C library) functions were pre-compiled and placed into the chip ROM. When the application is linked, we can use some of the functions from the chip ROM, reducing the total binary size. This also means that if we change something about how newlib should be compiled, we can no longer use some parts of newlib in ROM. In v5.0, we are changing the `time_t` type from 32-bit to a 64-bit one. Since the newlib in ROM was built for 32-bit time_t, we had to exclude some of the ROM functions from the linker script and start including the 64-bit versions of these functions into the app, instead. As a side effect, this brought a change to newlib's stdio layer. ESP32 ROM was built without `HAVE_BLKSIZE` feature, so it didn't check or use the block size reported by fstat function. We have enabled this feature for newer chips, but the ESP32 ROM was already built without HAVE_BLKSIZE. So it always assumed the default block size of 1024 bytes and the default buffer size of 128 bytes. Now, after we did the time_t change and replaced some ROM functions with the new ones built from source, HAVE_BLKSIZE feature got "enabled" for the ESP32. Since your littlefs port sets the block size in fstat function to 4096 bytes, this value is now taken into account.

Since some of the newer chips (like ESP32-C3) have newlib in ROM built with HAVE_BLKSIZE, the test also doesn't terminate if you run it on C3 with IDF v4.4!

Removing `st->st_blksize = efs->cfg.block_size;` line from `vfs_littlefs_fstat` makes newlib use the default 128 buffer size, and the test terminates correctly on ESP32 with IDF v5.0 and on ESP32-C3 with IDF v4.4.

Unfortunately I don't have more time to look into this issue today, but I will file an issue internally to check these findings and describe these details in the IDF Programming Guide and possibly in the release notes of v5.0.

Reply:
Another thing you could do is to keep the block size and modify the test to check the return code from `fclose`.

Reply:
Thank you for looking into it!

So for now, would we say that removing the `st->st_blksize = efs->cfg.block_size;` line is the correct fix? Or is this just kicking the problem further down? Is some inefficiency introduced?

Reply:
Essentially removing the block size setting will result in default buffer size (128 bytes) being used. That's the same behavior as for ESP32 until v5.0.

Having larger buffer sizes may improve performance, but this mostly depends on littlefs, it's a bit hard to predict. You can experiment with this by testing read/write performance and calling `setvbuf` to adjust the FILE* buffer size.
I know that this makes big difference for FAT on SD cards, but relatively small difference for FAT in SPI flash.

Reply:
Regardless of what you do about the block size, i suggest also fixing the test case to check fclose result.